src/clj/rems/db/email_outbox.clj

```diff
@@ -0,0 +1,99 @@
+(ns rems.db.email-outbox
+  (:require [clj-time.core :as time]
+            [clojure.test :refer [deftest is testing]]
+            [rems.db.core :as db]
+            [rems.db.pg-util :refer [pg-interval->joda-duration joda-duration->pg-interval]]
+            [rems.json :as json])
+  (:import [org.joda.time Period Duration DateTime]))
+
+(def ^Duration max-backoff (-> (Period.) (.withHours 12) (.toStandardDuration)))
+
+(defn put! [{:keys [email deadline]}]
+  (:id (db/put-to-email-outbox! {:email (json/generate-string email)
+                                 :deadline deadline})))
+
+(defn- fix-row-from-db [row]
+  {:email-outbox/id (:id row)
+   :email-outbox/email (json/parse-string (:email row))
+   :email-outbox/created (:created row)
+   :email-outbox/next-attempt (:next_attempt row)
+   :email-outbox/latest-attempt (:latest_attempt row)
+   :email-outbox/latest-error (:latest_error row)
+   :email-outbox/backoff (pg-interval->joda-duration (:backoff row))
+   :email-outbox/deadline (:deadline row)})
+
+(defn get-emails
+  ([]
+   (get-emails nil))
+  ([{:keys [ids due-now?]}]
+   (->> (db/get-email-outbox {:ids ids
+                              :due-now? due-now?})
+        (map fix-row-from-db))))
+
+(defn get-email-by-id [id]
+  (first (get-emails {:ids [id]})))
+
+(defn- next-attempt [email now error]
+  (let [^DateTime next-attempt (-> now (.plus (:email-outbox/backoff email)))
+        ^DateTime deadline (:email-outbox/deadline email)
+        ^Duration backoff (-> (:email-outbox/backoff email) (.multipliedBy 2))]
+    (assoc email
+           :email-outbox/latest-attempt now
+           :email-outbox/latest-error error
+           :email-outbox/next-attempt (if (-> next-attempt (.isAfter deadline))
+                                        nil
+                                        next-attempt)
+           :email-outbox/backoff (if (-> backoff (.isLongerThan max-backoff))
+                                   max-backoff
+                                   backoff))))
+
+(deftest test-next-attempt
+  (let [now (DateTime. 1000)]
+    (testing "basic case"
+      (is (= {:email-outbox/latest-attempt now
+              :email-outbox/latest-error "the error"
+              :email-outbox/next-attempt (DateTime. 3000) ; now + backoff
+              :email-outbox/backoff (Duration. 4000) ; 2 * backoff
+              :email-outbox/deadline (DateTime. 60000000)
+              :unrelated-keys "should be kept"}
+             (next-attempt {:email-outbox/backoff (Duration. 2000)
+                            :email-outbox/deadline (DateTime. 60000000)
+                            :unrelated-keys "should be kept"}
+                           now "the error"))))
+
+    (testing "max backoff reached"
+      (is (= {:email-outbox/latest-attempt now
+              :email-outbox/latest-error "the error"
+              :email-outbox/next-attempt (.plus now max-backoff)
+              :email-outbox/backoff max-backoff
```

Comment:
I'm interested in the test case where backoff is `(- max-backoff "1 second")` or something like that